schema/labels.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2025 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schema
+
+import (
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+const (
+	// Special label names and selectors for schema.Metadata fields.
+	// They are currently private to ensure __name__, __type__ and __unit__ are used
+	// together and remain extensible in Prometheus. See NewMetadataFromLabels and Metadata
+	// methods for the interactions with the labels package structs.
+	metricName = "__name__"
+	metricType = "__type__"
+	metricUnit = "__unit__"
+)
+
+// IsMetadataLabel returns true if the given label name is a special
+// schema Metadata label.
+func IsMetadataLabel(name string) bool {
+	return name == metricName || name == metricType || name == metricUnit
+}
+
+// Metadata represents the core metric schema/metadata elements that:
+// * are describing and identifying the metric schema/shape (e.g. name, type and unit).
+// * are contributing to the general metric/series identity.
+// * with the type-and-unit feature, are stored as Prometheus labels.
+//
+// Historically, similar information was encoded in the labels.MetricName (suffixes)
+// and in the separate metadata.Metadata structures. However, with the
+// type-and-unit-label feature (PROM-39), this information can be now stored directly
+// in the special schema metadata labels, which offers better reliability (e.g. atomicity),
+// compatibility and, in many cases, efficiency.
+//
+// NOTE: Metadata in the current form is generally similar (yet different) to:
+//   - The MetricFamily definition in OpenMetrics (https://prometheus.io/docs/specs/om/open_metrics_spec/#metricfamily).
+//     However, there is a small and important distinction around the metric name semantics
+//     for the "classic" representation of complex metrics like histograms. The
+//     Metadata.Name follows the __name__ semantics. See Name for details.
+//   - Original metadata.Metadata entries. However, not all fields in that metadata
+//     are "identifiable", notably the help field, plus metadata does not contain Name.
+type Metadata struct {
```

Comment:
```suggestion
type Identity struct {
```

I think we'll create confusion if we use the same name for the traditional metadata, even though, as you said, they are different. I'm giving one name suggestion based on what you're adding in the comment:

```go
// * are describing and identifying the metric schema/shape (e.g. name, type and unit).
// * are contributing to the general metric/series identity.
```

but of course, open to other suggestions

Reply:
Yea,  my aim to address confusion was because it's a schema metadata not just metadata. We will use exactly the same information to correlate and link to schemas eventually (and optionally). The additional metadata naming advantage is that this particular struct IS PLANNED to replace existing WAL metadata records (at least for type and unit).

`Schema Identity` is not too bad option as well. Although "Schema ID" is something else, there will be literally ID for each metric in the schema, so there will be some confusion. Plus if you look on usage, it's very clear what `metricMeta` is about, it will be odd to see in parsers someone creating and storing `schema ID` for metric without schema 🤔 

I'm also open for suggestion, but we might have a situation where there's no good option. Also it's internal naming so we can change anytime.

Any other suggestion on naming here e.g. @beorn7 ? 

We could also name it `NameTypeUnit` for now (: The struct is useful for  grouping helper methods.

Reply:
In other words, what's more readable?

```
	if !ev.enableDelayedNameRemoval && dropName {
		metric = metric.DropReserved(schema.IsMetadataLabel)
	}
```

or 

```
	if !ev.enableDelayedNameRemoval && dropName {
		metric = metric.DropReserved(schema.IsIdentityLabel)
	}
```

Reply:
At this moment I prefer IsIdentityLabel, but I'm not strongly opinionated about it :) It's totally ok to keep as it is if you prefer

Reply:
No strong opinion from my side. I see "metadata" as a very fluffy term by now that might mean anything and nothing. So maybe it's good to use here because we don't need to be overly precise or specific.

Every label is part of the metric's identity, so "identity label" sounds more misleading…

Reply:
Yea, it's tricky. I would say let's stick to metadata for now and if we have more voices that prefer @ArthurSens option or anything else we can quickly change later.

Thanks!

Reply:
slight preference for IsMetadataLabel.  Maybe IsSchemaIdentity label could make sense, but at first glance IsIdentityLabel is confusing since all labels are identifying.